File: naikan-client/src/app/project/bom-detail.ts
import {
  Commit,
  Contact,
  Developer,
  Documentation,
  Environment,
  Integration,
  License,
  Organization,
  Project,
  Team,
  Technology
} from "@naikan/shared";

export interface BomDetail {
  id: string;
  timestamp?: Date;
  project?: Project;
  organization?: Organization;
  environments?: Environment[];
  teams?: Team[];
  developers?: Developer[];
  contacts?: Contact[];
  technologies?: Technology[];
  licenses?: License[];
  documentations?: Documentation[];
  integrations?: Integration[];
  tags?: string[];
  repository: Repository;
}

export interface Repository {
  name?: string;
  url?: string;
  firstCommit?: Commit;
  totalCommits?: number;
  defaultBranch?: string;
  branchesCount?: number;
  tagsCount?: number;
}

File: naikan-server/src/main/java/com/enofex/naikan/project/BomDetail.java
package com.enofex.naikan.project;

import com.enofex.naikan.model.Commit;
import com.enofex.naikan.model.Contacts;
import com.enofex.naikan.model.Developers;
import com.enofex.naikan.model.Documentations;
import com.enofex.naikan.model.Environments;
import com.enofex.naikan.model.Integrations;
import com.enofex.naikan.model.Licenses;
import com.enofex.naikan.model.Organization;
import com.enofex.naikan.model.Project;
import com.enofex.naikan.model.Tags;
import com.enofex.naikan.model.Teams;
import com.enofex.naikan.model.Technologies;
import java.time.LocalDateTime;

record BomDetail(String id, LocalDateTime timestamp,
                        Project project, Organization organization, Environments environments,
                        Teams teams, Developers developers, Contacts contacts,
                        Technologies technologies, Licenses licenses,
                        Documentations documentations, Integrations integrations, Tags tags,
                        Repository repository) {

  BomDetail {
    environments = environments != null ? environments : Environments.empty();
    teams = teams != null ? teams : Teams.empty();
    developers = developers != null ? developers : Developers.empty();
    contacts = contacts != null ? contacts : Contacts.empty();
    technologies = technologies != null ? technologies : Technologies.empty();
    licenses = licenses != null ? licenses : Licenses.empty();
    documentations = documentations != null ? documentations : Documentations.empty();
    integrations = integrations != null ? integrations : Integrations.empty();
    tags = tags != null ? tags : Tags.empty();
  }

  record Repository(String name, String url, Commit firstCommit, int totalCommits,
                           String defaultBranch, int branchesCount, int tagsCount) {

  }
}
